File: data_acquisition_shadow.cpp
//
// このコードは Hsin-Yuan Huang (https://momohuang.github.io/)
// によって作成されました。 詳細は以下の論文を参照してください:
//  "Predicting Many Properties of a Quantum System from Very Few Measurements"
//  (非常に少ない測定から量子系の多くの特性を予測する)
//
#include <algorithm>
#include <cassert>
#include <cmath>
#include <ctime>
#include <fstream>
#include <iostream>
#include <sstream>
#include <stdio.h>
#include <string.h>
#include <string>
#include <utility>
#include <vector>

using namespace std;
const int INF = 999999999; // 無限大として扱う非常に大きな数

double eta = 0.9; // チューニングが必要なハイパーパラメータ

int system_size;
int number_of_observables;
int number_of_measurements_per_observable;
int max_k_local;

//
// 以下の関数はファイル: observable_file_name を読み込み、
// [observables] と [observables_acting_on_ith_qubit] を更新します。
//
vector<vector<pair<int, int>>> observables; // 予測する観測量 (Pauli observable)
vector<vector<vector<int>>> observables_acting_on_ith_qubit;
vector<double> observables_weight;

void read_all_observables(char *observable_file_name) {
  ifstream observable_fstream;
  observable_fstream.open(observable_file_name, ifstream::in);

  if (observable_fstream.fail()) {
    fprintf(stderr,
            "\n====\nError: 入力ファイル \"%s\" が存在しません。\n====\n",
            observable_file_name);
    exit(-1);
  }

  // システムサイズを読み込む
  observable_fstream >> system_size;

  // 以下の数値を初期化 (後で変更される)
  max_k_local = 0;
  number_of_observables = 0;

  //
  // observables_acting_on_ith_qubit の構造:
  //   observables_acting_on_ith_qubit[ith_qubit][0 or 1 or 2]
  //     ith_qubit に対して X (0), Y (1), Z (2)
  //     を適用する観測量のインデックスリストを返します
  //
  observables_acting_on_ith_qubit.clear();
  vector<int> single_list;
  vector<vector<int>> pauli_list;
  pauli_list.push_back(single_list);
  pauli_list.push_back(single_list);
  pauli_list.push_back(single_list);
  for (int i = 0; i < system_size; i++) {
    observables_acting_on_ith_qubit.push_back(pauli_list);
  }

  // 局所観測量を行ごとに読み込む
  string line;
  int observable_counter = 0;
  while (getline(observable_fstream, line)) {
    if (line == "\n")
      continue;
    istringstream single_line_stream(line);

    int k_local;
    single_line_stream >> k_local;
    max_k_local = max(max_k_local, k_local);

    vector<pair<int, int>> ith_observable;

    for (int k = 0; k < k_local; k++) {
      char pauli_observable[5];
      int position_of_pauli;
      single_line_stream >> pauli_observable >> position_of_pauli;

      assert(pauli_observable[0] == 'X' || pauli_observable[0] == 'Y' ||
             pauli_observable[0] == 'Z');

      int pauli_encoding = pauli_observable[0] - 'X'; // X -> 0, Y -> 1, Z -> 2

      observables_acting_on_ith_qubit[position_of_pauli][pauli_encoding]
          .push_back(observable_counter);
      ith_observable.push_back(make_pair(position_of_pauli, pauli_encoding));
    }

    double weight;
    int X = single_line_stream.rdbuf()->in_avail();
    if (X == 0)
      weight = 1.0;
    else
      single_line_stream >> weight;

    observables_weight.push_back(weight);

    observables.push_back(ith_observable);
    observable_counter++;
  }
  number_of_observables = observable_counter;
  observable_fstream.close();

  return;
}

//
// 以下の関数はこのプログラムの使用法を表示します。
//
void print_usage() {
  fprintf(stderr, "使用法:\n");
  fprintf(
      stderr,
      "./shadow_data_acquisition -d [観測量ごとの測定回数] [observable.txt]\n");
  fprintf(
      stderr,
      "    これは古典シャドウの非ランダム化 (derandomized) バージョンです。\n");
  fprintf(stderr, "    [observable.txt] にあるすべての観測量を少なくとも "
                  "[観測量ごとの測定回数] 回測定するための\n");
  fprintf(stderr, "    パウリ測定のリストを出力します。\n");
  fprintf(stderr, "<または>\n");
  fprintf(stderr,
          "./shadow_data_acquisition -r [総測定回数] [システムサイズ]\n");
  fprintf(stderr,
          "    これは古典シャドウのランダム化 (randomized) バージョンです。\n");
  fprintf(stderr, "    指定された [システムサイズ] に対して、合計 [総測定回数] "
                  "回の繰り返しのための\n");
  fprintf(stderr, "    パウリ測定のリストを出力します。\n");
  return;
}

//
// 以下の関数は乗法重み更新 (multiplicative weight update) 法を実行します。
// これは古典シャドウのランダムなパウリ測定を非ランダム化するために使用されます。
//
vector<double>
    log1ppow1o3k; // log1ppow1o3k[k] = log(1 + (e^(-eta / 2) - 1) / 3^k)
double sum_log_value = 0.0;
int sum_cnt = 0.0;
double
fail_prob_pessimistic(int cur_num_of_measurements, int how_many_pauli_to_match,
                      double weight,
                      double shift) { // "悲観的推定による失敗確率" を意味します
  double log1pp0 =
      (how_many_pauli_to_match < INF ? log1ppow1o3k[how_many_pauli_to_match]
                                     : 0.0);

  if (floor(weight * number_of_measurements_per_observable) <=
      cur_num_of_measurements)
    return 0;

  double log_value = -eta / 2 * cur_num_of_measurements + log1pp0;
  sum_log_value += (log_value / weight);
  sum_cnt++;
  return 2 * exp((log_value / weight) - shift);
}

int main(int argc, char *argv[]) {
  if (argc != 4) {
    print_usage();
    return -1;
  }

  //
  // 古典シャドウのランダム化バージョンを実行
  //
  if (strcmp(argv[1], "-r") == 0) {
    //
    // この実行のためのランダムシードを設定
    //
    srand(time(NULL));

    //
    // パラメータを読み込む
    //
    system_size = stoi(argv[3]);
    int number_of_total_measurements = stoi(argv[2]);
    char Pauli[] = {'X', 'Y', 'Z'};

    //
    // 古典シャドウのランダム化バージョン
    // ランダムにパウリ基底 (X, Y, Z) を選択して測定します。
    //
    for (int i = 0; i < number_of_total_measurements; i++) {
      for (int j = 0; j < system_size; j++) {
        printf("%c ", Pauli[rand() % 3]);
      }
      printf("\n");
    }
  }
  //
  // 古典シャドウの非ランダム化バージョンを実行
  //
  else if (strcmp(argv[1], "-d") == 0) {
    read_all_observables(argv[3]);

    //
    // 非ランダム化プロセスの効率的な使用のためにいくつかの定数を事前計算
    //
    double expm1eta = expm1(-eta / 2); // expm1eta = e^(-eta / 2) - 1
    for (int k = 0; k < max_k_local + 1; k++) {
      log1ppow1o3k.push_back(log1p(pow(1.0 / 3.0, k) * expm1eta));
    }

    //
    // 各局所観測量をこれだけの回数測定したい
    //
    number_of_measurements_per_observable = stoi(argv[2]);

    //
    // 古典シャドウの非ランダム化バージョン
    //
    // ランダムに選ぶ代わりに、未測定の観測量を効率的にカバーできるような
    // パウリ基底を決定論的に (貪欲法的に) 選択します。
    //

    // すべての観測量について、
    // 以前のすべての測定繰り返しの中で、その観測量が何回測定されたか
    vector<int> cur_num_of_measurements; // "現在の測定回数" を意味します
    cur_num_of_measurements.resize(number_of_observables, 0); // 0 で初期化

    // すべての観測量について、
    // 現在の測定繰り返しにおいて、その観測量を測定するために
    // いくつのパウリ演算子が一致する必要があるか
    vector<int> how_many_pauli_to_match;
    how_many_pauli_to_match.resize(number_of_observables);

    for (int measurement_repetition = 0; measurement_repetition < INF;
         measurement_repetition++) {
      for (int i = 0; i < (int)observables.size(); i++)
        how_many_pauli_to_match[i] =
            observables[i].size(); // k-local 観測量の場合は k で初期化

      double shift = (sum_cnt == 0) ? 0 : sum_log_value / sum_cnt;
      sum_log_value = 0.0;
      sum_cnt = 0;

      for (int ith_qubit = 0; ith_qubit < system_size; ith_qubit++) {
        double prob_of_failure[3]; // X, Y, または Z を選ぶための失敗確率
        double smallest_prob_of_failure = -1;

        //
        // 現在の繰り返しで ith_qubit に対してパウリ測定を選ぶ場合
        //
        for (int pauli = 0; pauli < 3; pauli++) {
          prob_of_failure[pauli] = 0;

          // すべてのパウリ観測量 p について、スコアを計算できる
          for (int p = 0; p < 3; p++) {
            for (int i : observables_acting_on_ith_qubit[ith_qubit][p]) {
              if (pauli == p) {
                int pauli_to_match_next_step =
                    how_many_pauli_to_match[i] == INF
                        ? INF
                        : how_many_pauli_to_match[i] - 1;
                double prob_next_step = fail_prob_pessimistic(
                    cur_num_of_measurements[i], pauli_to_match_next_step,
                    observables_weight[i], shift);
                double prob_current_step = fail_prob_pessimistic(
                    cur_num_of_measurements[i], how_many_pauli_to_match[i],
                    observables_weight[i], shift);
                prob_of_failure[pauli] += prob_next_step - prob_current_step;
              } else {
                double prob_next_step =
                    fail_prob_pessimistic(cur_num_of_measurements[i], INF,
                                          observables_weight[i], shift);
                double prob_current_step = fail_prob_pessimistic(
                    cur_num_of_measurements[i], how_many_pauli_to_match[i],
                    observables_weight[i], shift);
                prob_of_failure[pauli] += prob_next_step - prob_current_step;
              }
            }
          }

          if (smallest_prob_of_failure == -1)
            smallest_prob_of_failure = prob_of_failure[pauli];
          else
            smallest_prob_of_failure =
                min(smallest_prob_of_failure, prob_of_failure[pauli]);
        }

        // 最も低い失敗確率を持つものを選ぶ
        int the_best_pauli = 0;
        for (int pauli = 0; pauli < 3; pauli++) {
          if (smallest_prob_of_failure == prob_of_failure[pauli]) {
            printf("%c ", 'X' + pauli);
            the_best_pauli = pauli;
            break;
          }
        }

        for (int pauli = 0; pauli <= 2; pauli++) {
          for (int i : observables_acting_on_ith_qubit[ith_qubit][pauli]) {
            if (the_best_pauli == pauli) {
              if (how_many_pauli_to_match[i] != INF)
                how_many_pauli_to_match[i] -= 1;
            } else
              how_many_pauli_to_match[i] = INF;
          }
        }
      }
      printf("\n");

      for (int i = 0; i < (int)observables.size(); i++)
        if (how_many_pauli_to_match[i] == 0)
          cur_num_of_measurements[i]++;

      //
      // すべての観測量の測定回数をチェック
      //
      int success = 0;
      for (int i = 0; i < (int)observables.size(); i++)
        if (cur_num_of_measurements[i] >=
            floor(observables_weight[i] *
                  number_of_measurements_per_observable))
          success += 1;
      fprintf(stderr, "[Status %d: %d]\n", measurement_repetition + 1, success);

      if (success == (int)observables.size())
        break;
    }
  }
}

File: data_acquisition_shadow.py
#
# このコードは Hsin-Yuan Huang (https://momohuang.github.io/) によって作成されました。
# 詳細は以下の論文を参照してください:
#  "Predicting Many Properties of a Quantum System from Very Few Measurements"
#  (非常に少ない測定から量子系の多くの特性を予測する)
# この Python バージョンは C++ バージョンよりも低速です。(コードの最適化が少ないため)
# しかし、理解しやすく、拡張もしやすいはずです。
#
import sys
import random
import math

def randomized_classical_shadow(num_total_measurements, system_size):
    #
    # 古典シャドウのランダム化バージョンの実装
    #
    #    num_total_measurements: int, 測定ラウンドの総数 (T)
    #    system_size: int, 量子系の量子ビット数 (n)
    #
    measurement_procedure = []
    for t in range(num_total_measurements):
        # 各量子ビットに対してランダムにパウリ基底 (X, Y, Z) を選択
        single_round_measurement = [random.choice(["X", "Y", "Z"]) for i in range(system_size)]
        measurement_procedure.append(single_round_measurement)
    return measurement_procedure

def derandomized_classical_shadow(all_observables, num_of_measurements_per_observable, system_size, weight=None):
    #
    # 古典シャドウの非ランダム化バージョンの実装
    #
    #     all_observables: パウリ観測量のリスト。各パウリ観測量はタプルのリストです。
    #                       ("X", position) または ("Y", position) または ("Z", position) の形式
    #     num_of_measurements_per_observable: int, 各観測量に対する測定回数
    #     system_size: int, 量子系の量子ビット数
    #     weight: None または 各観測量に対する係数のリスト
    #             None -- このパラメータを無視
    #             リスト -- 対応する重みで各観測量の測定回数を変更
    #
    if weight is None:
        weight = [1.0] * len(all_observables)
    assert(len(weight) == len(all_observables))

    sum_log_value = 0
    sum_cnt = 0

    def cost_function(num_of_measurements_so_far, num_of_matches_needed_in_this_round, shift = 0):
        eta = 0.9 # 変更可能なハイパーパラメータ
        nu = 1 - math.exp(-eta / 2)

        nonlocal sum_log_value
        nonlocal sum_cnt

        cost = 0
        for i, zipitem in enumerate(zip(num_of_measurements_so_far, num_of_matches_needed_in_this_round)):
            measurement_so_far, matches_needed = zipitem
            # 既に十分な回数測定されている場合はスキップ
            if num_of_measurements_so_far[i] >= math.floor(weight[i] * num_of_measurements_per_observable):
                continue

            if system_size < matches_needed:
                V = eta / 2 * measurement_so_far
            else:
                V = eta / 2 * measurement_so_far - math.log(1 - nu / (3 ** matches_needed))
            cost += math.exp(-V / weight[i] - shift)

            sum_log_value += V / weight[i]
            sum_cnt += 1

        return cost

    def match_up(qubit_i, dice_roll_pauli, single_observable):
        for pauli, pos in single_observable:
            if pos != qubit_i:
                continue
            else:
                if pauli != dice_roll_pauli:
                    return -1 # 不一致
                else:
                    return 1 # 一致
        return 0 # 無関係

    num_of_measurements_so_far = [0] * len(all_observables)
    measurement_procedure = []

    for repetition in range(num_of_measurements_per_observable * len(all_observables)):
        # "system_size" 個の量子ビットに対する並列測定の1ラウンド
        num_of_matches_needed_in_this_round = [len(P) for P in all_observables]
        single_round_measurement = []

        shift = sum_log_value / sum_cnt if sum_cnt > 0 else 0;
        sum_log_value = 0.0
        sum_cnt = 0

        for qubit_i in range(system_size):
            cost_of_outcomes = dict([("X", 0), ("Y", 0), ("Z", 0)])

            for dice_roll_pauli in ["X", "Y", "Z"]:
                # 測定結果が "dice_roll_pauli" になると仮定
                for i, single_observable in enumerate(all_observables):
                    result = match_up(qubit_i, dice_roll_pauli, single_observable)
                    if result == -1:
                        num_of_matches_needed_in_this_round[i] += 100 * (system_size+10) # 測定不可能
                    if result == 1:
                        num_of_matches_needed_in_this_round[i] -= 1 # Pauli X/Y/Z が1つ一致

                cost_of_outcomes[dice_roll_pauli] = cost_function(num_of_measurements_so_far, num_of_matches_needed_in_this_round, shift=shift)

                # 仮定を元に戻す
                for i, single_observable in enumerate(all_observables):
                    result = match_up(qubit_i, dice_roll_pauli, single_observable)
                    if result == -1:
                        num_of_matches_needed_in_this_round[i] -= 100 * (system_size+10) # 測定不可能
                    if result == 1:
                        num_of_matches_needed_in_this_round[i] += 1 # Pauli X/Y/Z が1つ一致

            for dice_roll_pauli in ["X", "Y", "Z"]:
                if min(cost_of_outcomes.values()) < cost_of_outcomes[dice_roll_pauli]:
                    continue
                # コストが最小の測定基底をここで選択
                single_round_measurement.append(dice_roll_pauli)
                for i, single_observable in enumerate(all_observables):
                    result = match_up(qubit_i, dice_roll_pauli, single_observable)
                    if result == -1:
                        num_of_matches_needed_in_this_round[i] += 100 * (system_size+10) # 測定不可能
                    if result == 1:
                        num_of_matches_needed_in_this_round[i] -= 1 # Pauli X/Y/Z が1つ一致
                break

        measurement_procedure.append(single_round_measurement)

        for i, single_observable in enumerate(all_observables):
            if num_of_matches_needed_in_this_round[i] == 0: # すべての量子ビットの測定が完了
                num_of_measurements_so_far[i] += 1

        success = 0
        for i, single_observable in enumerate(all_observables):
            if num_of_measurements_so_far[i] >= math.floor(weight[i] * num_of_measurements_per_observable):
                success += 1

        if success == len(all_observables):
            break

    return measurement_procedure

#
# 以下のコードは、コマンドラインインターフェースを介してこのコードを実行する場合にのみ使用されます
#
if __name__ == "__main__":
    def print_usage():
        print("Usage:\n", file=sys.stderr)
        print("python3 data_acquisition_shadow -d [観測量ごとの測定回数] [observable.txt]", file=sys.stderr)
        print("    これは古典シャドウの非ランダム化 (derandomized) バージョンです。", file=sys.stderr)
        print("    [observable.txt] にあるすべての観測量を少なくとも [観測量ごとの測定回数] 回測定するための", file=sys.stderr)
        print("    パウリ測定のリストを出力します。", file=sys.stderr)
        print("<or>\n", file=sys.stderr)
        print("python3 data_acquisition_shadow -r [総測定回数] [システムサイズ]", file=sys.stderr)
        print("    これは古典シャドウのランダム化 (randomized) バージョンです。", file=sys.stderr)
        print("    指定された [システムサイズ] に対して、合計 [総測定回数] 回の繰り返しのための", file=sys.stderr)
        print("    パウリ測定のリストを出力します。", file=sys.stderr)

    if len(sys.argv) != 4:
        print_usage()
    if sys.argv[1] == "-d":
        with open(sys.argv[3]) as f:
            content = f.readlines()
        system_size = int(content[0])

        all_observables = []
        for line in content[1:]:
            one_observable = []
            for pauli_XYZ, position in zip(line.split(" ")[1::2], line.split(" ")[2::2]):
                one_observable.append((pauli_XYZ, int(position)))
            all_observables.append(one_observable)
        measurement_procedure = derandomized_classical_shadow(all_observables, int(sys.argv[2]), system_size)
    elif sys.argv[1] == "-r":
        measurement_procedure = randomized_classical_shadow(int(sys.argv[2]), int(sys.argv[3]))
    else:
        print_usage()

    for single_round_measurement in measurement_procedure:
        print(" ".join(single_round_measurement))
